### PR TITLE
modify bridging nft to be soulbound and levels based

### PR DESCRIPTION
**Pull request by ETeissonniere**
- **change meta**
- **introduce notion of levels and different token URIs**
- **move to a soulbound token**
- **remove unused batch**
- **holder limitation still allows for level ups**
- **integrate levels**
- **update deploy script**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -110,17 +110,24 @@ forge script script/DeployNodlMigration.sol --zksync --rpc-url https://sepolia.e
...
+The `MigrationNFT` contract allows the minting of a reward soulbound NFT when users bridge enough tokens through `NODLMigration`. Users can "level up" depending on the amount of tokens they bring in, with each levels being a sorted list of bridged amounts.
+
+You will need to set the following environment variables:
 - `N_MIGRATION`: address of the `NODLMigration` contract
-- `N_MAX_NFTS`: maximum number of NFTs available
-- `N_MIN_AMOUNT`: minimum amount of tokens to bridge to be elligible for a NFT.
-- `N_TOKENS_URI`: URI of NFT metadata.
+- `N_MAX_HOLDERS`: maximum number of participants
+- `N_TOKENS_URI_ROOT`: URI of NFT metadata folder where each level metadata is later presented
+- `N_LEVELS`: number of levels to set
```
Comment (aliXsed) [inline README.md:119]: What are these levels?

File: README.md [hunk 2]
```diff
@@ -110,17 +110,24 @@ forge script script/DeployNodlMigration.sol --zksync --rpc-url https://sepolia.e
...
 N_MIGRATION=0x1427d38B967435a3F8f476Cda0bc4F51fe66AF4D \
-N_MAX_NFTS=10000 \
-N_MIN_AMOUNT=100000000000000000000 \
-N_TOKENS_URI="ipfs://Qmcy3noasf25Z9rJkg818rrRaP6mShca8cHtfBfdf2VVJJ" \
+N_MAX_HOLDERS=10000 \
+N_TOKENS_URI_ROOT="ipfs://Qmcy3noasf25Z9rJkg818rrRaP6mShca8cHtfBfdf2VVJJ" \
+N_LEVELS=3 \
+N_LEVELS_0=100 \
+N_LEVELS_1=200 \
+N_LEVELS_2=300 \
```
Comment (aliXsed) [inline README.md:130]: It seems not very hard to have them as 
`N_LEVELS=100,200,300`
then read it through vm.envString and parse it to an array. Just a nice to have suggestion.

File: src/bridge/MigrationNFT.sol [hunk 3]
```diff
@@ -3,38 +3,51 @@
...
+import {Strings} from "openzeppelin-contracts/contracts/utils/Strings.sol";
 import {NODLMigration} from "./NODLMigration.sol";
 
 contract MigrationNFT is ERC721 {
+    using Strings for uint256;
+
     uint256 public nextTokenId;
-    uint256 public maxNFTs;
-    uint256 public minAmount;
+    uint256 public maxHolders;
```
Comment (aliXsed) [inline src/bridge/MigrationNFT.sol:13]:
```suggestion
    uint256 public immutable maxHolders;
```

File: src/bridge/MigrationNFT.sol [hunk 4]
```diff
@@ -3,38 +3,51 @@
...
     NODLMigration public migration;
 
-    string internal tokensURI;
+    string internal tokensURIRoot;
+
+    uint256[] public levels;
```
Comment (aliXsed) [inline src/bridge/MigrationNFT.sol:18]:
```suggestion
    string internal immutable tokensURIRoot;

    uint256[] public immutable levels;
```

File: src/bridge/MigrationNFT.sol [hunk 5]
```diff
@@ -3,38 +3,51 @@
...
 import {NODLMigration} from "./NODLMigration.sol";
 
 contract MigrationNFT is ERC721 {
+    using Strings for uint256;
+
     uint256 public nextTokenId;
-    uint256 public maxNFTs;
-    uint256 public minAmount;
+    uint256 public maxHolders;
     NODLMigration public migration;
```
Comment (aliXsed) [inline src/bridge/MigrationNFT.sol:14]:
```suggestion
    NODLMigration public immutable migration;
```

File: src/bridge/MigrationNFT.sol [hunk 6]
```diff
@@ -3,38 +3,51 @@
...
 
-    error TooManyNFTs();
+    error TooManyHolders();
     error AlreadyClaimed();
+    error NoLevelUp();
     error ProposalDoesNotExist();
-    error UnderMinimumAmount();
     error NotExecuted();
-    error AlreadyAHolder();
+    error Soulbound();
```
Comment (aliXsed) [inline src/bridge/MigrationNFT.sol:30]: Perhaps a bit better naming for this error could be as:
```suggestion
    error SoulboundTokenTransferAttempt();
```

File: src/bridge/MigrationNFT.sol [hunk 7]
```diff
@@ -93,28 +120,32 @@ contract MigrationNFT is ERC721 {
...
+        alreadyHolder = balanceOf(target) > 0;
+        if (!alreadyHolder && individualHolders == maxHolders) {
+            revert TooManyHolders();
         }
     }
+
+    function _update(address to, uint256 tokenId, address auth) internal override(ERC721) returns (address) {
+        address from = _ownerOf(tokenId);
+        if (from != address(0) && to != address(0)) {
+            // not a burn or mint, since this is a soulbound token we need to revert
```
Comment (aliXsed) [inline src/bridge/MigrationNFT.sol:145]:
```suggestion
            // only burn or mint is allowed for a soulbound token
```

File: src/bridge/MigrationNFT.sol [hunk 8]
```diff
@@ -3,38 +3,51 @@
...
         migration = _migration;
-        maxNFTs = _maxNFTs;
-        minAmount = _minAmount;
-        tokensURI = _tokensURI;
+        maxHolders = _maxHolders;
+        tokensURIRoot = _tokensURIRoot;
+        levels = _levels;
+
+        for (uint256 i = 1; i < levels.length; i++) {
+            assert(levels[i] > levels[i - 1]);
```
Comment (aliXsed) [inline src/bridge/MigrationNFT.sol:49]: I suggest using a more expressive error message here. Something like:
```suggestion
            if (levels[i] <= levels[i - 1]) {
               revert LevelsNotStrictlyIncreasing();
            }
```

File: README.md (hunk 1 above)
Comment (ETeissonniere) [reply README.md:119]: Basically depending on how many tokens you bridge you get better looking NFTs. Idea from products.

File: README.md (hunk 2 above)
Comment (ETeissonniere) [reply README.md:130]: I agree but as far as I know this not quite doable without implementing string manipulation functions or using a library. Let me know if you can think of a straightforward way to do it

File: README.md (hunk 1 above)
Comment (aliXsed) [reply README.md:119]: So perhaps it's better to be called `N_AMOUNT_REUQIRED_FOR_RANK_0` or `N_AMOUNTS_REQUIRED_FOR_RANKS`

File: README.md (hunk 2 above)
Comment (aliXsed) [reply README.md:130]: I also think you would need to manipulate the string which is (processing wise) not bad for a deployment script

File: src/bridge/MigrationNFT.sol [hunk 9]
```diff
@@ -43,46 +56,60 @@ contract MigrationNFT is ERC721 {
...
+        // We effectively iterate over all the levels the `target` has YET
+        // to qualify for. This expressively skips levels the `target` has
+        // already qualified for.
+        uint256 currentLevel = holderToLevel[target];
+        for (uint256 i = currentLevel; i < levels.length; i++) {
+            if (amount >= levels[i]) {
+                levelsToMint[i - currentLevel] = i;
+                nbLevelsToMint++;
+            }
+        }
```
Comment (aliXsed) [inline src/bridge/MigrationNFT.sol:109]: If I'm reading the logic correctly here, an individual would be able to get multiples of tokens of the same rank if they have multiple bridge transactions of the same level. If I'm correct is that what we want or we want to issue only one token for an achieved rank?

Comment (aliXsed) [reply src/bridge/MigrationNFT.sol:109]: I think my own answer: `holderToLevel` keeps track of a person's currently expected next level! ;)

Comment (aliXsed) [reply src/bridge/MigrationNFT.sol:109]: In the light of my answer above I suggest naming `holderToLevel` as `holderNextLevel` and `currentLevel` as `nextLevel`. This means the loop starts from checking a person's eligibility from their next level up.

File: src/bridge/MigrationNFT.sol (hunk 3 above)
Comment (aliXsed) [reply src/bridge/MigrationNFT.sol:13]: Is it possible that we quickly issue all the NFTs for the lowest rank/level before we issue any high rank NFT?

Comment (ETeissonniere) [reply src/bridge/MigrationNFT.sol:13]: Yes that's a possibility, this is not supposed to be open forever but is instead a short lived incentive program

File: src/bridge/MigrationNFT.sol (hunk 9 above)
Comment (ETeissonniere) [reply src/bridge/MigrationNFT.sol:109]: Yes that's fine, makes it more expressive. And yes indeed we just allocate all soulbounds if you are on overachiever

File: src/bridge/MigrationNFT.sol (hunk 4 above)
Comment (ETeissonniere) [reply src/bridge/MigrationNFT.sol:18]: FYI @aliXsed you cannot make `uint256[]` and `string` immutable. `solc` will complain via `Error (6377): Immutable variables cannot have a non-value type.`

File: src/bridge/MigrationNFT.sol (hunk 9 above)
Comment (ETeissonniere) [reply src/bridge/MigrationNFT.sol:109]: Did some renaming and added documentation comments to clarify

File: src/bridge/MigrationNFT.sol (hunk 8 above)
Comment (ETeissonniere) [reply src/bridge/MigrationNFT.sol:49]: I was lazy :D. Will update

**Pull request merged**